File: test/cpp/measureSpaceUsage.cpp
#include <iostream>
#include <cstdlib>

#include <sstream>
#include <memory>

#include <hoytech/error.h>
#include <hoytech/hex.h>

#include "negentropy.h"
#include "negentropy/storage/BTreeLMDB.h"
#include "negentropy/storage/BTreeMem.h"
#include "negentropy/storage/btree/debug.h"
#include "negentropy/storage/Vector.h"






int main() {
    auto env = lmdb::env::create();
    env.set_max_dbs(64);
    env.set_mapsize(1'000'000'000ULL);
    env.open("testdb/", 0);


    negentropy::storage::BTreeLMDB btree;

    {
        auto txn = lmdb::txn::begin(env);
        btree.setup(txn, "negentropy");
        txn.commit();
    }


    {
        auto txn = lmdb::txn::begin(env);

        btree.withWriteTxn(txn, [&]{
            auto add = [&](uint64_t timestamp){
                negentropy::Item item(timestamp, std::string(32, '\x01'));
                btree.insert(item);
            };

            for (size_t i = 1; i < 100'000; i++) add(i);
        });

        txn.commit();
    }

    {
        auto txn = lmdb::txn::begin(env, 0, MDB_RDONLY);

        auto cursor = lmdb::cursor::open(txn, btree.dbi);

        std::string_view key, val;
        size_t minStart = negentropy::MAX_U64;
        size_t maxEnd = 0;

        if (cursor.get(key, val, MDB_FIRST)) {
            do {
                size_t ptrStart = (size_t)val.data();
                size_t ptrEnd = ptrStart + sizeof(negentropy::storage::btree::Node);
                if (ptrStart < minStart) minStart = ptrStart;
                if (ptrEnd > maxEnd) maxEnd = ptrEnd;
            } while (cursor.get(key, val, MDB_NEXT));
        }

        std::cout << "data," << negentropy::storage::btree::MAX_ITEMS << "," << sizeof(negentropy::storage::btree::Node) << "," << (maxEnd - minStart) << std::endl;
    }

    return 0;
}

File: test/cpp/measureSpaceUsage.pl
system(qq{ perl -pi -e 's/MIN_ITEMS = \\d+/MIN_ITEMS = 2/' ../../cpp/negentropy/storage/btree/core.h });
system(qq{ perl -pi -e 's/REBALANCE_THRESHOLD = \\d+/REBALANCE_THRESHOLD = 4/' ../../cpp/negentropy/storage/btree/core.h });
system(qq{ perl -pi -e 's/MAX_ITEMS = \\d+/MAX_ITEMS = 6/' ../../cpp/negentropy/storage/btree/core.h });

for (my $i = 6; $i < 128; $i += 2) {
    print "DOING ITER $i\n";
    system(qq{ perl -pi -e 's/MAX_ITEMS = \\d+/MAX_ITEMS = $i/' ../../cpp/negentropy/storage/btree/core.h });
    system("rm -f measureSpaceUsage && make measureSpaceUsage && rm -f testdb/data.mdb && ./measureSpaceUsage >> measureSpaceUsage.log");
}
